design/accepted/kv_auth.md

```diff
@@ -0,0 +1,115 @@
+- `Policies`s key will be in the form of `policies/<DisplayName>`.
+- `Credentials`s key will be in the form of `users/<UserName>/credentials/<AccessKeyID>`.
+
+### Handling Secondary Indexes
+
+Below are 2 options that vary in efficiency and complexity of the implementation.
+ 
+#### Store just the minimum indexes
+
+Keep only the minimum that is required to represent the relationship of the entities:
+
+- A `User` membership of a `Group` under `groups/<DisplayName>/users/<Username>`.
+- A `Policy` attached to a `User` under `users/<Username>/policies/<DisplayName>`.
+- A `Policy` attached to a `Group` under `groups/<Displayname>/policies/<DisplayName>`.
+
+Any deletion of an entity would first remove all its secondary indexes, then the entity itself.
+For example, deleting a `Policy` would have to list all `User`s & `Group`'s `Policies` and delete any attachment if existed.
+Only then it should delete the `Policy` entity from the store.
+
+Listing by anything other than the key will require to list all the entities which are relevant.
+Some examples:
+
+1. Finding a `User` by `AccessKeyID` will require to list all the users.
+1. Listing the `User`'s effective policies will require to list all entities under `groups/`.
+
+The amount of entities in the Auth world isn't big (<10k) and cached in the server, it's unlikely it will incur a 
+notable performance degradation.
+
+#### Store all secondary indexes
+
+In addition to the indexes in the above suggestion, also store:
+
+- A `User` membership of a `Group` under `users/<Username>/groups/<DisplayName>`.
+- `Credentials` attached to a `User` under `credentials/<AccessKeyID>/<Username>`.
+- A `Policy` attached to a `User` under `policies/<DisplayName>/users/<Username>`.
+- A `Policy` attached to a `Group` under `policies/<DisplayName>/groups/<DisplayName>`.
+
+Every operation that updates a relationship between two entities would have to be stored everywhere.
+The upside is a possible increase in performance ("possible" since we read less entities, but perform much more round-trips to the cache/store).
+
+### Working with no locks
+
+In the Postgres era, we rely on it to keep the entities and relationship consistent.
+For example, a deleted `Policy` would cascade to the `auth_user_policies` & `auth_group_policies` tables.
+
+
+In the KV world, we must cleanup the secondary indexes first, before deleting the entity itself.
+Since the KV is lock-free, we might still get into troubles.
+For example, a `Policy` is requested to be deleted. The operation starts by listing all `User`s & `Group`s with that `Policy` attached to it.
+While iterating, it's possible that another unlisted `User` would be attached to the `Policy`. Once the `Policy` is deleted,
+we would remain with an attachment to it. There are at least 2 mitigations for it:
+
+1. Do (a) list and delete of secondary indexes (b) delete the entity (c) another round of list and delete.
```

Comment:
In the case why not perform the deletion order in reverse? delete entity -> delete secondary indexes without making another round?

Reply:
@N-o-Z this might cause an issue as now there are users attached to a policy that doesn't exist, I think we should check how the user's effective policy generation would behave if that's the case.

Reply:
Assuming this is an issue, we will still encounter it in the second round of list and delete

Reply:
By design, we don't have a DB lock. Hence with with every implementation we choose, extra care must be taken.
When reading secondary index, one must make sure that the principal entities exist. If not, it's ok to delete the secondary index.